src/lib.rs

```diff
@@ -428,89 +503,161 @@ impl<Types: SwarmTypes> Future for IpfsFuture<Types> {
         // begin by polling the swarm so that initially it'll first have chance to bind listeners
         // and such. TODO: this no longer needs to be a swarm event but perhaps we should
         // consolidate logging of these events here, if necessary?
+
+        let mut done = false;
+
         loop {
-            let inner = {
-                let next = self.swarm.next_event();
-                futures::pin_mut!(next);
-                match next.poll(ctx) {
-                    Poll::Ready(inner) => inner,
-                    Poll::Pending => break,
+            loop {
+                let inner = {
+                    let next = self.swarm.next_event();
+                    futures::pin_mut!(next);
+                    match next.poll(ctx) {
+                        Poll::Ready(inner) => inner,
+                        Poll::Pending if done => return Poll::Pending,
+                        Poll::Pending => break,
```

Comment:
so it's now equivalent to the original logic where the swarm was polled like this:
```rust
loop {
    loop { poll ipfs event }
    poll swarm
}
```

Reply:
so I updated the loop to fix this in the ipfs-rfc pr.

```rust
fn poll(mut self: Pin<&mut Self>, ctx: &mut Context) -> Poll<Self::Output> {
        loop {
            let swarm_event = match Pin::new(&mut self.swarm).poll_next(ctx) {
                Poll::Ready(Some(event)) => Some(event),
                Poll::Ready(None) => return Poll::Ready(()),
                Poll::Pending => None,
            };

            let block_store_event = match Pin::new(&mut self.block_store).poll_next(ctx) {
                Poll::Ready(Some(event)) => Some(event),
                Poll::Ready(None) => return Poll::Ready(()),
                Poll::Pending => None,
            };

            let ipfs_event = match Pin::new(&mut self.ipfs_events).poll_next(ctx) {
                Poll::Ready(Some(event)) => Some(event),
                Poll::Ready(None) => return Poll::Ready(()),
                Poll::Pending => None,
            };

            if swarm_event.is_none() && block_store_event.is_none() && ipfs_event.is_none() {
                return Poll::Pending;
            }

            if let Some(event) = swarm_event {
            ....
```

Reply:
I don't think it's equivalent to original logic where ipfs event is polled first, because if you try look at the `connect_two_nodes` integration test, the identity will be served before swarm has gotten the network up (explained also in the comment first thing inside the background task). If anything, I think that we should aim to do less per poll call and schedule wakeup for the remaining work...

Reply:
The integration test passes

Reply:
Ah yes my original comment was wrong